app/gamestate.ts

```diff
@@ -0,0 +1,390 @@
+        if (robot.energy < 0) {
+            robot.energy += wire.Costs.moves.shield;
+            robot.setResult({type: 'shield', success: false});
+            this.hold(robot);
+            return true;
+        }
+        robot.shielded = true;
+        robot.setResult({type: 'shield', success: true});
+        this.entities.robots.forEach(bot => {
+            if (bot.inRange(robot, wire.ViewDistance)) {
+                bot.addObservation({type: 'shield', location: robot.location});
+            }
+        });
+        return false;
+    }
+
+    private pendingMoveTargets: {[index: string]: {bots: RobotGameObject[], location: wire.Point}} = {};
+    move(robot: RobotGameObject, action: wire.MoveActionMessage): boolean {
+        if (!action.arguments) {
+            this.hold(robot);
+            robot.setResult({type: 'move', success: false});
+            return true;
+        }
+        const direction = action.arguments.direction;
+        if (typeof direction !== 'number') {
+            this.hold(robot);
+            robot.setResult({type: 'move', success: false});
+            return true;
+        }
+        const location = directionToLoaction(robot.location, direction);
+        if (0 > location.x || location.x > (this.width - 1)
+         || 0 > location.y || location.y > (this.height - 1)) {
+            this.hold(robot);
+            robot.setResult({type: 'move', success: false, obstacle: wire.ObstacleType.Wall});
+            return true;
+        }
+        const key = `${location.x},${location.y}`;
+        this.pendingMoveTargets[key] = this.pendingMoveTargets[key] || {location, bots: []};
+        this.pendingMoveTargets[key].bots.push(robot);
+        return false;
+    }
+
+    commitPendingMoves() {
+        // Collect a list of all move resolutions (and how to undo them)
+        const undo: {[index: string]: {func: () => void, bot: RobotGameObject}} = {};
+        for (const key in this.pendingMoveTargets) {
+            const {location, bots: [bot]} = this.pendingMoveTargets[key];
+            const oldLocation = {x: bot.location.x, y: bot.location.y};
+            bot.location.x = location.x;
+            bot.location.y = location.y;
+            undo[bot.name] = {func: ((bot: RobotGameObject, oldLocation: wire.Point) => () => {
+                bot.location.x = oldLocation.x;
+                bot.location.y = oldLocation.y;
+            })(bot, oldLocation), bot};
+        }
+        // Collect all robots at each location
+        const contents: {[index: string]: {bots: RobotGameObject[], location: wire.Point}} = {};
+        this.entities.robots.forEach(bot => {
+            const key = `${bot.location.x},${bot.location.y}`;
+            contents[key] = contents[key] || {location: {x: bot.location.x, y: bot.location.y}, bots: []};
+            contents[key].bots.push(bot);
+        });
+        function undoMovement(undoFunc: () => void, bot: RobotGameObject) {
```

Comment:
The algorithm needs to be recursive, so unless I want to rewrite it to use a stack/queue, there will be a function involved. IMO, the recursive algorithm is simpler.